### ISSUE TITLE
Using `legend(labelcolor="markerfacecolor")` with a scatter plot throws an error

### ISSUE THREAD

**Issue by agude** [matplotlib/matplotlib#20577]
### Bug report

**`legend(labelcolor="markerfacecolor")` raises error when used with `scatter()`**

Adding a legend with colored text using any of `linecolor`, `markerfacecolor`, or `markeredgecolor`  as the values for the argument `labelcolor` with a scatter plot results in a `TypeError: unhashable type: 'numpy.ndarray'`.

**Code for reproduction**

```python
import matplotlib.pyplot as plt
import numpy as np

fig, ax = plt.subplots()

ax.scatter(np.random.rand(500)-0.65, np.random.rand(500), label="First dataset")
ax.scatter(np.random.rand(500)-0.35, np.random.rand(500), label="Second dataset")

ax.legend(labelcolor="linecolor")

fig.savefig("/tmp/test.png")
```

**Actual outcome**

```
Traceback (most recent call last):
  File "test.py", line 16, in <module>
    fig.savefig("/tmp/test.png")
  File "/home/agude/bin/anaconda3/lib/python3.8/site-packages/matplotlib/figure.py", line 3005, in savefig
    self.canvas.print_figure(fname, **kwargs)
  File "/home/agude/bin/anaconda3/lib/python3.8/site-packages/matplotlib/backends/backend_qt5agg.py", line 78, in print_figure
    super().print_figure(*args, **kwargs)
  File "/home/agude/bin/anaconda3/lib/python3.8/site-packages/matplotlib/backend_bases.py", line 2255, in print_figure
    result = print_method(
  File "/home/agude/bin/anaconda3/lib/python3.8/site-packages/matplotlib/backend_bases.py", line 1669, in wrapper
    return func(*args, **kwargs)
  File "/home/agude/bin/anaconda3/lib/python3.8/site-packages/matplotlib/backends/backend_agg.py", line 508, in print_png
    FigureCanvasAgg.draw(self)
  File "/home/agude/bin/anaconda3/lib/python3.8/site-packages/matplotlib/backends/backend_agg.py", line 406, in draw
    self.figure.draw(self.renderer)
  File "/home/agude/bin/anaconda3/lib/python3.8/site-packages/matplotlib/artist.py", line 74, in draw_wrapper
    result = draw(artist, renderer, *args, **kwargs)
  File "/home/agude/bin/anaconda3/lib/python3.8/site-packages/matplotlib/artist.py", line 51, in draw_wrapper
    return draw(artist, renderer, *args, **kwargs)
  File "/home/agude/bin/anaconda3/lib/python3.8/site-packages/matplotlib/figure.py", line 2780, in draw
    mimage._draw_list_compositing_images(
  File "/home/agude/bin/anaconda3/lib/python3.8/site-packages/matplotlib/image.py", line 132, in _draw_list_compositing_images
    a.draw(renderer)
  File "/home/agude/bin/anaconda3/lib/python3.8/site-packages/matplotlib/artist.py", line 51, in draw_wrapper
    return draw(artist, renderer, *args, **kwargs)
  File "/home/agude/bin/anaconda3/lib/python3.8/site-packages/matplotlib/_api/deprecation.py", line 431, in wrapper
    return func(*inner_args, **inner_kwargs)
  File "/home/agude/bin/anaconda3/lib/python3.8/site-packages/matplotlib/axes/_base.py", line 2921, in draw
    mimage._draw_list_compositing_images(renderer, self, artists)
  File "/home/agude/bin/anaconda3/lib/python3.8/site-packages/matplotlib/image.py", line 132, in _draw_list_compositing_images
    a.draw(renderer)
  File "/home/agude/bin/anaconda3/lib/python3.8/site-packages/matplotlib/artist.py", line 51, in draw_wrapper
    return draw(artist, renderer, *args, **kwargs)
  File "/home/agude/bin/anaconda3/lib/python3.8/site-packages/matplotlib/legend.py", line 606, in draw
    bbox = self._legend_box.get_window_extent(renderer)
  File "/home/agude/bin/anaconda3/lib/python3.8/site-packages/matplotlib/offsetbox.py", line 352, in get_window_extent
    w, h, xd, yd, offsets = self.get_extent_offsets(renderer)
  File "/home/agude/bin/anaconda3/lib/python3.8/site-packages/matplotlib/offsetbox.py", line 438, in get_extent_offsets
    whd_list = [c.get_extent(renderer)
  File "/home/agude/bin/anaconda3/lib/python3.8/site-packages/matplotlib/offsetbox.py", line 438, in <listcomp>
    whd_list = [c.get_extent(renderer)
  File "/home/agude/bin/anaconda3/lib/python3.8/site-packages/matplotlib/offsetbox.py", line 347, in get_extent
    w, h, xd, yd, offsets = self.get_extent_offsets(renderer)
  File "/home/agude/bin/anaconda3/lib/python3.8/site-packages/matplotlib/offsetbox.py", line 474, in get_extent_offsets
    whd_list = [c.get_extent(renderer)
  File "/home/agude/bin/anaconda3/lib/python3.8/site-packages/matplotlib/offsetbox.py", line 474, in <listcomp>
    whd_list = [c.get_extent(renderer)
  File "/home/agude/bin/anaconda3/lib/python3.8/site-packages/matplotlib/offsetbox.py", line 347, in get_extent
    w, h, xd, yd, offsets = self.get_extent_offsets(renderer)
  File "/home/agude/bin/anaconda3/lib/python3.8/site-packages/matplotlib/offsetbox.py", line 438, in get_extent_offsets
    whd_list = [c.get_extent(renderer)
  File "/home/agude/bin/anaconda3/lib/python3.8/site-packages/matplotlib/offsetbox.py", line 438, in <listcomp>
    whd_list = [c.get_extent(renderer)
  File "/home/agude/bin/anaconda3/lib/python3.8/site-packages/matplotlib/offsetbox.py", line 347, in get_extent
    w, h, xd, yd, offsets = self.get_extent_offsets(renderer)
  File "/home/agude/bin/anaconda3/lib/python3.8/site-packages/matplotlib/offsetbox.py", line 474, in get_extent_offsets
    whd_list = [c.get_extent(renderer)
  File "/home/agude/bin/anaconda3/lib/python3.8/site-packages/matplotlib/offsetbox.py", line 474, in <listcomp>
    whd_list = [c.get_extent(renderer)
  File "/home/agude/bin/anaconda3/lib/python3.8/site-packages/matplotlib/offsetbox.py", line 823, in get_extent
    bbox, info, yd = self._text._get_layout(renderer)
  File "/home/agude/bin/anaconda3/lib/python3.8/site-packages/matplotlib/text.py", line 294, in _get_layout
    if key in self._cached:
TypeError: unhashable type: 'numpy.ndarray'
```

**Expected outcome**

The legend text should match the color of the points.

**Matplotlib version**

  * Operating system: Ubuntu Linux (`Linux hostname 5.4.0-77-generic #86~18.04.1-Ubuntu SMP Fri Jun 18 01:23:22 UTC 2021 x86_64 x86_64 x86_64 GNU/Linux`)
  * Matplotlib version (`import matplotlib; print(matplotlib.__version__)`): 3.4.2
  * Matplotlib backend (`print(matplotlib.get_backend())`): Qt5Agg
  * Python version: Python 3.8.10
  * Other libraries: Numpy 1.20.2

Matplotlib is currently installed with `pip install matplotlib==3.4.2` in a Conda install to force the most recent version... Yeah I know.

**Comment by jklymak** [comment]
Did this ever work?

**Comment by agude** [comment]
> Did this ever work?

I don't think so. I think this is the bug, which is in the original PR:

### Debugging

I think the bug is here: https://github.com/matplotlib/matplotlib/blob/master/lib/matplotlib/legend.py#L538

`get_color()` called on a `Line2D` returns a color `str`, like `'#1f77b4'`.

But `get_facecolor()` (or `get_edgecolor`) called on a `PathCollection` (which is what `scatter()` generates) returns a `numpy.ndarray` of shape `(1, 4)`, like `array([[1.        , 0.49803922, 0.05490196, 1.        ]])`

I think to fix it we need to do something like `color = get_facecolor()[0]`.

**Comment by jklymak** [comment]
Right.  The whole point of `scatter` is that you can color the markers independently, so `markerfacecolor` is a 2-D array.   It probably shouldn't error out like this, but its hard to know what the "right" thing to do is.

**Comment by agude** [comment]
> ... but its hard to know what the "right" thing to do is.

I'd argue we should copy the behavior of the handle color. 

For a single color, handle color uses that, for a color map... Well for that I'd have to look deeper. :-/

**Comment by oscargus** [comment]
> for a color map

Emit a user warning that the first color in the list is selected? Still better than a crash.

**Comment by oscargus** [comment]
If anyone is interested in working on this, probably the solution is to add some logic between these two lines:
https://github.com/matplotlib/matplotlib/blob/028f07c869c4c0084c134a8b613aad4f5d111c7f/lib/matplotlib/legend.py#L548-L549

Basically, one would like to detect if color is not a single color value but a list of color values. If the length is 1 or all values are the same, just pick that value, if not emit a warning(?) and select the first.

Marking this as a good first issue, but medium since although the general approach is described, the details must be worked out. Tests can include the example above as a "smoke test", tests that the correct colors are selected for the legend, plus a test that the warning is correctly emitted in case there are multiple colors.

**Comment by chahak13** [comment]
Hi! I was wondering if I can pick up this issue?

**Comment by timhoffm** [comment]
Go for it. We don't formally assign issues.

**Comment by chahak13** [comment]
@timhoffm @oscargus I started working on this but I'm not sure about the behavior we want in the different cases, especially when having a cmap. If the length is 1 or the values are the same then I agree, we can just pick the first value of `color` and emit a warning, but if there's a cmap provided, then the first value that `get_facecolor` will return will be from the default color cycles and not from the cmap values, right? Can you confirm this? That's what I see on my end. I believe that a better solution could be that we emit a different warning when a cmap is used and select the value of `cmap(1.0)` as the color for the legend.

**Comment by timhoffm** [comment]
In the case of a multi-color scatter (either fixed-colors or colormapped, I suggest to fall back to the text color and not apply any artist color (in the face of ambiguity, refuse the temptation to guess). I'm a bit unsure whether we want to warn or do this silently. I tend towards silence here. There's a valid case for a legend containing a colormapped scatter plus some other artists. It can be desired to have all artists 'markerfacecolor'ed when possible and accept a the non-colored scatter.

**Comment by chahak13** [comment]
@timhoffm I see your point, and I think I agree. Having the text color for multi-color scatters would make more sense. I do want to point out that that's different than what this issue previously had discussed, wherein the idea was to just select the first color if there are multiple colors. I personally am leaning toward your suggestion of having just the text color for multi-colored but it's a design decision.

**Comment by timhoffm** [comment]
Let's go with text color.

**Comment by chahak13** [comment]
Sounds good. Thanks!

**Comment by QuLogic** [comment]
Fixed by #24122
